### PR TITLE
Allow to create subscriptions independently with topic

### LINKED ISSUE #71: Allow creating subscriptions for existing topics
Hello,

Currently, you are not able to create subscriptions if you set create_topic to false. This doesn't let you create subs separately from topics. This is useful when your topic is used by one application and is in a google project that is different from a google project that needs a sub for a different application. This is where is constraint is set for pull subs:

https://github.com/terraform-google-modules/terraform-google-pubsub/blob/master/main.tf#L178

Thanks,

### PR DESCRIPTION
**Pull request by onimsha**
Fixes #71

### CODE REVIEW

**Comment by comment-bot-dev** [comment]
Thanks for the PR! 🚀<br/>✅ Lint checks have passed.

File: main.tf [hunk 1]
```diff
@@ -34,19 +34,18 @@ resource "google_project_iam_member" "token_creator_binding" {
 }
 
 resource "google_pubsub_topic_iam_member" "push_topic_binding" {
-  for_each = var.create_topic ? { for i in var.push_subscriptions : i.name => i } : {}
-
-  project = var.project_id
-  topic   = lookup(each.value, "dead_letter_topic", "projects/${var.project_id}/topics/${var.topic}")
-  role    = "roles/pubsub.publisher"
-  member  = "serviceAccount:${local.pubsub_svc_account_email}"
+  for_each = { for i in var.push_subscriptions : i.name => i if lookup(i, "dead_letter_topic", null) != null }
```
Comment (morgante) [inline main.tf:37]: This PR seems to be doing several other things at the same time, I'm not sure what this change is intended for?

**Review by morgante — CHANGES_REQUESTED** [review]

File: main.tf (hunk 1 above)
Comment (onimsha) [reply main.tf:37]: Because now we can set the `create_topic = false`, while still able to create the subscription.   
This iam role is ONLY needed when we set the dead_letter_topic, so if the subscription does not have it, it should not be created.

**Comment by onimsha** [comment]
@googlebot I signed it!

**Review by morgante — CHANGES_REQUESTED** [review]
Thanks for the PR. I'm not sure we want to change this so that each subscription can have a different topic though, as I think that's *very* different from the original intent of the module.

If you do want to have this, all of the subscriptions should be attached to the same topic.

**Comment by onimsha** [comment]
> Thanks for the PR. I'm not sure we want to change this so that each subscription can have a different topic though, as I think that's _very_ different from the original intent of the module.
> 
> If you do want to have this, all of the subscriptions should be attached to the same topic.

Got it. Then how about this ?  

* Introduce a new variable, `existing_topic`, to use in the case `create_topic = false` . 
* Let all the subscriptions use the same topic, declared in `existing_topic` 
* Add another example to use the `existing_topic` by creating the topic in a project, while subscriptions in another project. 

If you think we can go with this approach, i will make the changes.

**Comment by morgante** [comment]
@onimsha I'm not sure I follow your use case.

If you want to attach subscriptions without creating the topic, why wouldn't you just set `create_topic = false`?

**Comment by onimsha** [comment]
> @onimsha I'm not sure I follow your use case.
> 
> 
> 
> If you want to attach subscriptions without creating the topic, why wouldn't you just set `create_topic = false`?

Because many resources in this module depend on this `for_each = var.create_topic` . One it's set to be false none of them will be created

**Comment by morgante** [comment]
> Because many resources in this module depend on this for_each = var.create_topic . One it's set to be false none of them will be created

I think we should update those resources to check a separate `create_subscriptions` variable instead. So, the new flow would be:

1. If `create_topic` and `create_subscriptions` are both `true` (default) it creates both.
2. If only `create_subscriptions` is true, only create subscriptions.
3. If only `create_topic` is true, only create topic.

**Comment by onimsha** [comment]
> > Because many resources in this module depend on this for_each = var.create_topic . One it's set to be false none of them will be created
> 
> I think we should update those resources to check a separate `create_subscriptions` variable instead. So, the new flow would be:
> 
> 1. If `create_topic` and `create_subscriptions` are both `true` (default) it creates both.
> 2. If only `create_subscriptions` is true, only create subscriptions.
> 3. If only `create_topic` is true, only create topic.

Changes pushed. Can you take a look ?

File: variables.tf [hunk 2]
```diff
@@ -22,6 +22,7 @@ variable "project_id" {
 variable "topic" {
   type        = string
   description = "The Pub/Sub topic name"
+  default     = "pubsub-topic"
```
Comment (morgante) [inline variables.tf:25]: Why did you add this default? We definitely shouldn't have this.

File: main.tf [hunk 3]
```diff
@@ -175,10 +175,10 @@ resource "google_pubsub_subscription" "push_subscriptions" {
 }
 
 resource "google_pubsub_subscription" "pull_subscriptions" {
-  for_each = var.create_topic ? { for i in var.pull_subscriptions : i.name => i } : {}
+  for_each = var.create_subscriptions ? { for i in var.pull_subscriptions : i.name => i } : {}
 
   name    = each.value.name
-  topic   = google_pubsub_topic.topic.0.name
+  topic   = var.create_topic ? google_pubsub_topic.topic.0.name : each.value.topic_name
```
Comment (morgante) [inline main.tf:181]: Please do not allow configuring separate topics on a per-subscription basis. At that point, there's no point to use the module at all.

All subscriptions for a single module instance must connect to the same topic.

File: examples/subscriptions_only/variables.tf [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+variable "topic_project" {
+  type        = string
+  description = "The project ID of the topic"
+}
+
+
+variable "topic_name" {
+  type        = string
+  description = "The name for the Pub/Sub topic"
```
Comment (morgante) [inline examples/subscriptions_only/variables.tf:30]: This should not be variable.

File: examples/subscriptions_only/variables.tf [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+variable "topic_labels" {
```
Comment (morgante) [inline examples/subscriptions_only/variables.tf:33]: Remove this.

**Review by morgante — CHANGES_REQUESTED** [review]

File: variables.tf (hunk 2 above)
Comment (onimsha) [reply variables.tf]: removed

File: main.tf (hunk 3 above)
Comment (onimsha) [reply main.tf]: changed to `var.topic`

File: examples/subscriptions_only/variables.tf (hunk 5 above)
Comment (onimsha) [reply examples/subscriptions_only/variables.tf]: removed

File: examples/subscriptions_only/variables.tf (hunk 4 above)
Comment (onimsha) [reply examples/subscriptions_only/variables.tf]: removed

**Review by morgante — APPROVED** [review]
